gradle-plugin/plugin/src/main/java/com/yelp/codegen/plugin/GenerateTaskConfiguration.kt

```diff
@@ -1,18 +1,44 @@
 package com.yelp.codegen.plugin
 
+import io.swagger.parser.SwaggerParser
 import java.io.File
+import javax.inject.Inject
 import org.gradle.api.Action
 import org.gradle.api.Project
+import org.gradle.api.file.RegularFileProperty
 
-open class GenerateTaskConfiguration(@Suppress("UNUSED_PARAMETER") project: Project) {
-    var platform: String? = null
-    var packageName: String? = null
-    var specName: String? = null
-    var specVersion: String? = null
-    lateinit var inputFile: File
-    var outputDir: File? = null
-    var extraFiles: File? = null
-    var features: FeatureConfiguration = FeatureConfiguration()
+abstract class GenerateTaskConfiguration @Inject constructor(project: Project) {
+    val objects = project.objects
+
+    val platform = objects.property(String::class.java).convention("kotlin")
+    val packageName = objects.property(String::class.java).convention("com.codegen.default")
+    val specName = objects.property(String::class.java).convention("defaultname")
+    abstract val inputFile: RegularFileProperty
+    val specVersion = objects.property(String::class.java).convention(project.provider {
+        readVersionFromSpecfile(inputFile.get().asFile)
+    })
+    val outputDir = objects.directoryProperty().convention(project.layout.buildDirectory.dir("gen").get())
+
+    val extraFiles = objects.directoryProperty()
+    val features: FeatureConfiguration = FeatureConfiguration(objects)
 
     fun features(action: Action<FeatureConfiguration>) = action.execute(features)
+
+    companion object {
+        private fun readVersionFromSpecfile(specFile: File): String {
+            val swaggerSpec = SwaggerParser().readWithInfo(specFile.absolutePath, listOf(), false).swagger
```

Comment:
Still not 100% sure that this method belongs here. This currently pulls in an `import io.swagger.parser.SwaggerParser` that I would love to avoid having in the `GenerateTaskConfiguration` file. How about having it as a top level function in a separate file (à la Util file)?

Reply:
I moved that to `VersionHelper`, couldn't think of a better name but happy to change it.

Doing this actually got me wondering why the version was needed as input in the first place if it's already present in the spec file ? What are the use cases to override that ? Can a specFile contain several versions of an api ?

Reply:
> why the version was needed as input in the first place if it's already present in the spec file 

Ideally you should not override the version in the gradle file and use the one provided in the spec file. However there are 2 scenarios that made us think that this could have been a case:

- Updating the generator code/template. Say that you bump the `swagger-gradle-codegen` version and want to re-publish an artifact generated from the updated generator with a different scenario.

- Semantic versioning. We thought about having the final version as a composition of the version of the generator + the version of the spec. So you could have something like `<generator-version>.<spec-version>.<patch>`. Moreover the spec version could be a timestamp (say `2020010101`) and you would probably want to have a way to beautify it.

Generally the idea of having a way to override the version felt useful.